src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperAttributes.cs

```diff
@@ -0,0 +1,150 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class TagHelperAttributes<TAttributeValue> : IList<TagHelperAttribute<TAttributeValue>>
+        where TAttributeValue : class
+    {
+        private readonly IList<TagHelperAttribute<TAttributeValue>> _attributes;
+
+        public TagHelperAttributes()
+        {
+            _attributes = new List<TagHelperAttribute<TAttributeValue>>();
+        }
+
+        public int Count => _attributes.Count;
+
+        public bool IsReadOnly => false;
+
+        public IEnumerable<string> Keys => _attributes.Select(attribute => attribute.Key);
+
+        public IEnumerable<TAttributeValue> Values => _attributes.Select(attribute => attribute.Value);
```

Comment:
we should separate responsibilities here.  rather than one class, should have two:
1) implement `IList<TagHelperAttribute<TAttributeValue>>` to maintain all attributes form the Razor source or just use a standard `List<T>`
2) implement `IDictionary<string, IList<TagHelperAttribute<TAttributeValue>>>` to support attribute lookup or just use a standard `Dictionary<TKey, TValue>`. (haven't looked far enough ahead to see these properties and methods used but they likely are)
 - this could be a projection over the `List<T>` above to avoid the two getting out of sync